### DISCUSSION TITLE
How to determine if egui didn't do anything with a mouse event?

### DISCUSSION THREAD

**Discussion by buzmeg** [emilk/egui#495] [Q&A]
If I click the mouse, how do I determine that egui didn't do anything with the click?

Thanks.

**Comment by parasyte** [comment]
I'm not entirely sure what you mean by `egui` not doing anything with a click. It will always do _something_ with a click if you tell it to sense click interactions and the user clicks on the widget.

The information you get back from `egui` is all in the [`Response`](https://docs.rs/egui/0.12.0/egui/struct.Response.html) type. Specifically, the composed `Sense` type will tell `egui` which interactions are expected, and the various methods on the `Response` can be used to check which (if any) interactions were performed by the user on the current frame.

Some widgets will do this internally without writing any extra code. For instance, the `RadioButton` widget will select itself when clicked (and deselect others in the same group). But these internal interactions are all provided with specific `Sense`s which may or may not ignore various interactions, depending on the widget. If you are asking about which of these internal interactions are ignored, you shouldn't have to worry about it, because you can always do your own interaction handling at a higher level.

**Reply by buzmeg** [reply]
Let's say I have a bunch of egui widgets around the left edge, but the click occurred in the middle of the underlying OS window.  Or, the egui widgets have transparency, and the click occurred in a transparent part.

So, there was a mouse click.  It didn't intersect any egui widgets and make any state change, so I need to handle it myself.

Or, I have something selectable underneath the egui widget, but I want to know that egui ate the event so I don't also grab the thing underneath when I shouldn't.

How do I know that a mouse click did or didn't intersect any egui widgets?

**Reply by gents83** [reply]
Did you already triied `ctx.is_using_pointer()` and `ctx.wants_keyboard_input()`?

**Reply by Taywee** [reply]
In my case (about 2 years later), [`is_pointer_over_area`](https://docs.rs/egui/0.22.0/egui/struct.Context.html#method.is_pointer_over_area) was the solution to my problem.  Thanks for pointing me that direction.

For more information, I'm using Bevy, and the systems can be arbitrarily reordered, so I can't just check a response, because my code might run between the input being read and the egui system.  I need to be able to tell if the click interacted with egui, and also if it will interact with egui but hasn't yet.

I'm mostly making this comment to mention the connection to Bevy, so that it's easier to find for somebody with the same problem in the future.

**Comment by DrOptix** [comment]
I'm also interested in this. I'm working on an egui integration for my own hobby game engine.

It is possible I'm doing something wrong, but what I noticed is this:

I have an `egui::Window` and I scroll over it. I need to make it possible for the event to not be propagated to the next layer which renders the scene.

**Comment by aleksander** [comment]
As @parasyte already said `Response` is what you are looking for:
```rust
let response = egui::Window::new("Window 1").show(egui_ctx, |ui| {
    ui.vertical(|ui| {
        ui.label("Label 1");
        ui.label("Label 2");
        ui.label("Label 3");
    });
}).unwrap();
if response.hovered() {
    println!("hovered!");
}
```
But `hovered` is not working while you are dragging the window or clicking to the window header, so it is more reliable to check mouse cursor is inside the window rectangle.
```rust
if response.rect.contains(mouse) {
    println!("hovered!");
}
```
